src/persist-client/src/batch.rs

```diff
@@ -328,47 +359,126 @@ where
 
-        if self.records.len() == 0 {
-            // Use the first record to attempt to pre-size the builder
-            // allocations.
-            let additional = usize::saturating_add(self.size_hint, 1);
-            self.records
-                .reserve(additional, self.key_buf.len(), self.val_buf.len());
-        }
-        self.records.push(((&self.key_buf, &self.val_buf), t, d));
-        self.num_updates += 1;
+        let size = {
+            let k = &self.key_buf[k_range.clone()];
+            let v = &self.val_buf[v_range.clone()];
+            ColumnarRecordsBuilder::columnar_record_size(k, v)
+        };
+        let (ts, diff) = (ts.clone(), diff.clone());
 
-        // If we've filled up a chunk of ColumnarRecords, flush it out now to
-        // blob storage to keep our memory usage capped.
         let mut part_written = false;
-        // TODO: we need to ensure batch parts don't exceed [crate::PersistConfig::blob_target_size].
-        // currently our logic undercounts batch part sizes by separately checking key and value
-        // size against the target size, rather than their sum. a template for how to do this lives
-        // in [crate::internal::compact::Compactor].
-        //
-        // Related issue: https://github.com/MaterializeInc/materialize/issues/14579
-        for part in self.records.take_filled() {
-            // TODO: This upper would ideally be `[self.max_ts+1]` but
-            // there's nothing that lets us increment a timestamp. An empty
-            // antichain is guaranteed to correctly bound the data in this
-            // part, but it doesn't really tell us anything. Figure out how
-            // to make a tighter bound, possibly by changing the part
-            // description to be an _inclusive_ upper.
-            let upper = Antichain::new();
-            let since = Antichain::from_elem(T::minimum());
-            self.parts.write(part, upper, since).await;
+        // if we've filled up a batch part, flush out to blob to keep our memory usage capped.
+        if size + self.current_part_total_bytes > self.blob_target_size {
+            // TODO: we currently consolidate a part when we've reached our target size of
+            // _unconsolidated_ updates. This means the part we actually write to Blob may
+            // be (substantially) smaller post-consolidation. We could alternatively create
+            // a `BTreeMap<(&[u8], &[u8], T), D>`, consolidate as we go, and only write out
+            // a part once the data referenced by the map reaches our target size.
+            self.flush_current_part().await;
             part_written = true;
         }
 
+        self.current_part_total_bytes += size;
+        self.current_part_key_bytes += k_range.len();
+        self.current_part_value_bytes += v_range.len();
+
         if part_written {
+            // NB: we flush our shared buf's contents to Blob storage when we've hit our
```

Comment:
this is to let us use `self.blob_target_size` as a hard cap? I don't think that's worth the complexity over unconditionally adding the update and flushing if the new size is >= self.blob_target_size the size of individual updates should be small compared to self.blob_target_size and also, there's all sorts of overhead in our columnar format that isn't taken into account in the `ColumnarRecordsBuilder::columnar_record_size` computation

Reply:
okayokay fine, I'll accept the limit as a soft cap 😄